samwise.js

```diff
@@ -1,10 +1,10 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?